### PR TITLE
[IOTDB-5168] Refactor traverse of AbstractTreeVisitor to FA-based traverse

### PR DESCRIPTION
**Pull request by MarcosZyk**
### Design
see https://apache-iotdb.feishu.cn/docx/DjrAdDM3woXcn1x5Phlcyveznqe

### Performance Test

Micro means the scale of Tree is micro, no more than 10 nodes. Massive means the scale of Tree is massive, up to 20w nodes.

Precise: precisePattern "root.db.a_0.d_0.s_0"
\*: oneLevelWildcardPattern "root.db.\*.\*.s_0"
Internal\*\*": internalMultiLevelWildcardPattern "root.\*\*.s_0"
Tail\*\*: tailMultiLevelWildcardPattern "root.\*\*"
Regex: regexPattern "root.db.a\*.d\*.s_0"

#### Before This PR
[Micro-Precise] Max: 129ms; Min: 82ms; Avg: 88ms
[Micro-\*] Max: 164ms; Min: 151ms; Avg: 154ms
[Micro-Internal\*\*] Max: 77ms; Min: 74ms; Avg: 75ms
[Micro-Tail\*\*] Max: 73ms; Min: 70ms; Avg: 71ms
[Micro-Regex] Max: 171ms; Min: 166ms; Avg: 168ms
[Massive-Precise] Max: 108ms; Min: 97ms; Avg: 100ms
[Massive-\*] Max: 166ms; Min: 161ms; Avg: 163ms
[Massive-Internal\*\*] Max: 122ms; Min: 114ms; Avg: 117ms
[Massive-Tail\*\*] Max: 113ms; Min: 106ms; Avg: 109ms
[Massive-Regex] Max: 175ms; Min: 168ms; Avg: 171ms

#### This PR
[Micro-Precise] Max: 129ms; Min: 77ms; Avg: 83ms
[Micro-\*] Max: 68ms; Min: 63ms; Avg: 65ms
[Micro-Internal\*\*] Max: 89ms; Min: 86ms; Avg: 87ms
[Micro-Tail\*\*] Max: 75ms; Min: 72ms; Avg: 72ms
[Micro-Regex] Max: 152ms; Min: 148ms; Avg: 150ms
[Massive-Precise] Max: 85ms; Min: 83ms; Avg: 84ms
[Massive-\*] Max: 90ms; Min: 86ms; Avg: 87ms
[Massive-Internal\*\*] Max: 122ms; Min: 112ms; Avg: 115ms
[Massive-Tail\*\*] Max: 108ms; Min: 102ms; Avg: 104ms
[Massive-Regex] Max: 119ms; Min: 114ms; Avg: 116ms

### CODE REVIEW

File: node-commons/src/main/java/org/apache/iotdb/commons/path/fa/match/IStateMatchInfo.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+  /**
+   * @param index the target state's index of matched order
+   * @return the index(th) matched state
+   */
+  IFAState getMatchedState(int index);
+
+  int getMatchedStateSize();
+
+  /** @return the value index, the id in FA Graph, of the source state */
+  int getSourceStateIndex();
```
Comment (Cpaulyz) [inline node-commons/src/main/java/org/apache/iotdb/commons/path/fa/match/IStateMatchInfo.java:54]: This is easily confused with the index in IFAState. 

```suggestion
  int getSourceStateOrdinal();
```

File: node-commons/src/main/java/org/apache/iotdb/commons/path/fa/match/IStateMatchInfo.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+  /** @param sourceStateIndex the value index, the id in FA Graph, of the source state, */
+  void setSourceStateIndex(int sourceStateIndex);
```
Comment (Cpaulyz) [inline node-commons/src/main/java/org/apache/iotdb/commons/path/fa/match/IStateMatchInfo.java:57]:
```suggestion
  void setSourceStateOrdinal((int sourceStateIndex);
```

File: node-commons/src/main/java/org/apache/iotdb/commons/path/fa/match/IStateMatchInfo.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+  IFAState getOneMatchedState();
+
+  void addMatchedState(IFAState state);
+
+  /**
+   * @param index the target state's index of matched order
+   * @return the index(th) matched state
+   */
+  IFAState getMatchedState(int index);
```
Comment (Cpaulyz) [inline node-commons/src/main/java/org/apache/iotdb/commons/path/fa/match/IStateMatchInfo.java:49]:
```suggestion
  IFAState getMatchedState(int ordinal);
```

File: node-commons/src/main/java/org/apache/iotdb/commons/path/fa/SimpleNFA.java [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+
+  // initial state of this NFA and the only transition from this state is "root"
+  private final SimpleNFAState initialState = new SimpleNFAState(-1);
+  // all states corresponding to raw pattern nodes, with an extra prefixMatch state
+  private final SimpleNFAState[] states;
+
+  // singletonMap, "root"
+  private final Map<String, IFATransition> initialTransition;
+  // the transition matches only specific value
+  private final Map<String, IFATransition>[] preciseMatchTransitionTable;
```
[the remark this answers is not in the document]
Comment (MarcosZyk) [reply node-commons/src/main/java/org/apache/iotdb/commons/path/fa/SimpleNFA.java]: Description added.

File: node-commons/src/main/java/org/apache/iotdb/commons/path/fa/match/MatchedStateSet.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+      System.arraycopy(existingState, 0, array, 0, end);
+      existingState = array;
+    }
+    existingState[end++] = state.getIndex();
+    stateStatus[state.getIndex()] = true;
+  }
+
+  int getStateValueIndex(int matchIndex) {
+    return existingState[matchIndex];
+  }
```
[the remark this answers is not in the document]
Comment (MarcosZyk) [reply node-commons/src/main/java/org/apache/iotdb/commons/path/fa/match/MatchedStateSet.java]: Excellent suggestion! Already fixed.

File: node-commons/src/main/java/org/apache/iotdb/commons/path/fa/match/IStateMatchInfo.java (hunk 1 above)
Comment (MarcosZyk) [reply node-commons/src/main/java/org/apache/iotdb/commons/path/fa/match/IStateMatchInfo.java]: Fixed.

File: node-commons/src/main/java/org/apache/iotdb/commons/path/fa/match/IStateMatchInfo.java (hunk 2 above)
Comment (MarcosZyk) [reply node-commons/src/main/java/org/apache/iotdb/commons/path/fa/match/IStateMatchInfo.java]: Fixed.

File: node-commons/src/main/java/org/apache/iotdb/commons/path/fa/match/IStateMatchInfo.java (hunk 3 above)
Comment (MarcosZyk) [reply node-commons/src/main/java/org/apache/iotdb/commons/path/fa/match/IStateMatchInfo.java]: Fixed.

**Review by JackieTien97 — CHANGES_REQUESTED** [review]

File: node-commons/src/main/java/org/apache/iotdb/commons/path/fa/IFATransition.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iotdb.commons.path.fa;
+
+public interface IFATransition {
```
[the remark this answers is not in the document]
Comment (MarcosZyk) [reply node-commons/src/main/java/org/apache/iotdb/commons/path/fa/IFATransition.java:23]: Added.

File: node-commons/src/main/java/org/apache/iotdb/commons/path/fa/SimpleNFA.java [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+ *
+ * <p>initial -(root)-> state[0] -(sg)-> state[1] -(*)-> state[2] -(s)-> state[3] <br>
+ * state[3] is final state
+ *
+ * <p>Given path pattern root.**.d.s, the SimpleNFA is:
+ *
+ * <p>initial -(root)-> state[0] -(*)-> state[1] -(d)-> state[2] -(s)-> state[3] <br>
+ * with extra: state[2] -(*)-> state[1] <br>
+ * state[3] is final state
+ *
```
[the remark this answers is not in the document]
Comment (MarcosZyk) [reply node-commons/src/main/java/org/apache/iotdb/commons/path/fa/SimpleNFA.java]: Fixed.

File: node-commons/src/main/java/org/apache/iotdb/commons/path/fa/SimpleNFA.java [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+      }
+      if (nodes[patternIndex].equals(MULTI_LEVEL_PATH_WILDCARD)) {
+        return true;
+      }
+      if (nodes[patternIndex].equals(ONE_LEVEL_PATH_WILDCARD)) {
+        return true;
+      }
+      if (nodes[patternIndex].contains(ONE_LEVEL_PATH_WILDCARD)) {
+        return regexPatternMap.get(patternIndex).matcher(event).matches();
+      }
```
[the remark this answers is not in the document]
Comment (MarcosZyk) [reply node-commons/src/main/java/org/apache/iotdb/commons/path/fa/SimpleNFA.java]: Wonderful suggestion! Already implemented.

File: node-commons/src/main/java/org/apache/iotdb/commons/path/fa/SimpleNFA.java [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+      return nodes[patternIndex].equals(event);
+    }
+
+    @Override
+    public boolean isBatch() {
```
[the remark this answers is not in the document]
Comment (MarcosZyk) [reply node-commons/src/main/java/org/apache/iotdb/commons/path/fa/SimpleNFA.java]: Done.

File: node-commons/src/main/java/org/apache/iotdb/commons/path/fa/IFATransition.java [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
+  String getValue();
+
+  boolean isMatch(String event);
+
+  boolean isBatch();
```
[the remark this answers is not in the document]
Comment (MarcosZyk) [reply node-commons/src/main/java/org/apache/iotdb/commons/path/fa/IFATransition.java]: Done.

File: node-commons/src/main/java/org/apache/iotdb/commons/schema/tree/AbstractTreeVisitor.java [hunk 11]
```diff
@@ -487,54 +257,377 @@ protected String[] generateFullPathNodes(N node) {
...
+            }
+            if (matchedState == null) {
+              currentStateMatchInfo.setSourceStateOrdinal(sourceStateOrdinal - 1);
+              currentStateMatchInfo.setSourceTransitionIterator(transitionIterator);
+              currentNodeIndex--;
+              continue;
+            }
+          }
+
+          currentStateMatchInfo.addMatchedState(matchedState);
```
[the remark this answers is not in the document]
Comment (MarcosZyk) [reply node-commons/src/main/java/org/apache/iotdb/commons/schema/tree/AbstractTreeVisitor.java:597]: This change is not right. There's still trial in if block to find matchedState.

[the remark this answers is not in the document]
Comment (MarcosZyk) [reply node-commons/src/main/java/org/apache/iotdb/commons/schema/tree/AbstractTreeVisitor.java:597]: This change is not right. There's still trial in if block to find matchedState.

File: node-commons/src/main/java/org/apache/iotdb/commons/path/fa/match/IStateMatchInfo.java [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+
+  boolean hasOnlyPreciseMatchTransition();
+
+  boolean hasNoPreciseMatchTransition();
+
+  boolean isSingleFuzzyMatchTransition();
+
+  IFAState getOneMatchedState();
+
+  void addMatchedState(IFAState state);
```
[the remark this answers is not in the document]
Comment (MarcosZyk) [reply node-commons/src/main/java/org/apache/iotdb/commons/path/fa/match/IStateMatchInfo.java]: Added.

**Review by JackieTien97 — APPROVED** [review]

File: node-commons/src/main/java/org/apache/iotdb/commons/path/fa/match/MatchedStateSet.java [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+  private final boolean[] stateStatus;
+
+  private int[] existingState = new int[INITIAL_SIZE];
+
+  private int end = 0;
+
+  MatchedStateSet(int stateSize) {
+    stateStatus = new boolean[stateSize];
+  }
+
```
Comment (Cpaulyz) [inline node-commons/src/main/java/org/apache/iotdb/commons/path/fa/match/MatchedStateSet.java:37]: Just a suggestion for readability~

```suggestion
  /**
   * Construct an empty set with specified initial capacity.
   *
   * @param initialCapacity generally, we use stateSize as initialCapacity 
   */
  MatchedStateSet(int initialCapacity) {
    stateStatus = new boolean[initialCapacity];
  }
```

File: node-commons/src/main/java/org/apache/iotdb/commons/schema/tree/AbstractTreeVisitor.java [hunk 14]
```diff
@@ -487,54 +259,374 @@ protected String[] generateFullPathNodes(N node) {
...
+
+        saveResult(child, stateMatchInfo);
+        return;
+      }
     }
 
-    boolean isMatched(int index) {
-      return matchStatus[matchedIndex - index] == 1;
+    private Iterator<IFATransition> tryGetNextMatchedState(
+        N child, IFAState sourceState, IStateMatchInfo currentStateMatchInfo) {
```
[the remark this answers is not in the document]
Comment (MarcosZyk) [reply node-commons/src/main/java/org/apache/iotdb/commons/schema/tree/AbstractTreeVisitor.java:499]: Added.

File: node-commons/src/main/java/org/apache/iotdb/commons/schema/tree/AbstractTreeVisitor.java [hunk 15]
```diff
@@ -487,54 +259,374 @@ protected String[] generateFullPathNodes(N node) {
...
+            }
+          }
+
+          if (matchedState == null) {
+            while (++sourceStateOrdinal < parentStateMatchInfo.getMatchedStateSize()) {
+              sourceState = parentStateMatchInfo.getMatchedState(sourceStateOrdinal);
+              matchedStateSize = currentStateMatchInfo.getMatchedStateSize();
+              transitionIterator =
+                  tryGetNextMatchedState(currentNode, sourceState, currentStateMatchInfo);
+              if (matchedStateSize != currentStateMatchInfo.getMatchedStateSize()) {
```
[the remark this answers is not in the document]
Comment (MarcosZyk) [reply node-commons/src/main/java/org/apache/iotdb/commons/schema/tree/AbstractTreeVisitor.java:589]: Added.

**Pull request merged**